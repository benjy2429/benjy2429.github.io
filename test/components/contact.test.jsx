import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Contact from '../../src/components/contact';

const validState = {
  name: { value: 'testName', errors: [] },
  email: { value: 'testEmail', errors: [] },
  message: { value: 'testMessage', errors: [] }
};

const mockEvent = { preventDefault: () => {} };

describe('Contact component', () => {
  const originalFetch = window.fetch;
  const originalLocationAssign = window.location.assign;

  beforeEach(() => {
    window.fetch = jest
      .fn()
      .mockImplementation(() =>
        Promise.resolve({ json: () => Promise.resolve({ status: 200 }) })
      );
    window.location.assign = jest.fn();
  });

  afterAll(() => {
    window.fetch = originalFetch;
    window.location.assign = originalLocationAssign;
  });

  it('renders correctly', () => {
    const component = shallow(<Contact />);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('with validation errors', () => {
    it('renders validation errors on submit', async () => {
      const component = shallow(<Contact />);
      await component.instance().submitForm(mockEvent);
      component.update();

      const validationErrors = component.find('.validation-error');
      expect(validationErrors.length).toBeGreaterThan(0);
      expect(validationErrors.first().text()).toEqual(
        'Please enter a valid name'
      );
    });
  });

  describe('with a successful response', () => {
    it('redirects to the success page', () => {
      const component = shallow(<Contact />);
      component.setState(validState);
      component
        .instance()
        .submitForm(mockEvent)
        .then(() => {
          expect(window.location.assign).toHaveBeenCalledWith('/thanks');
        });
    });
  });

  describe('with a successful response', () => {
    it('redirects to the error page', () => {
      window.fetch.mockImplementation(() =>
        Promise.resolve({ json: () => Promise.resolve({ status: 500 }) })
      );

      const component = shallow(<Contact />);
      component.setState(validState);
      component
        .instance()
        .submitForm(mockEvent)
        .then(() => {
          expect(window.location.assign).toHaveBeenCalledWith('/formerror');
        });
    });
  });

  describe('with a failed response', () => {
    it('redirects to the error page', () => {
      window.fetch.mockImplementation(() => Promise.reject());

      const component = shallow(<Contact />);
      component.setState(validState);
      component
        .instance()
        .submitForm(mockEvent)
        .then(() => {
          expect(window.location.assign).toHaveBeenCalledWith('/formerror');
        });
    });
  });
});
