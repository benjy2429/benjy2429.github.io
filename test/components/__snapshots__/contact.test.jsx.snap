// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component renders correctly 1`] = `
<section
  className="contact"
>
  <div
    className="container"
  >
    <h2
      id="contact"
    >
      Contact
    </h2>
    <p>
      I hope Iâ€™ve covered everything but if you have any questions or comments, drop me a message.
    </p>
    <form
      onSubmit={[Function]}
    >
      <input
        name="honeypot"
        onChange={[Function]}
        type="hidden"
      />
      <label
        htmlFor="contact-name"
      >
        Name
      </label>
      <input
        className=""
        id="contact-name"
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <label
        htmlFor="contact-email"
      >
        Email
      </label>
      <input
        className=""
        id="contact-email"
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
      />
      <label
        htmlFor="contact-message"
      >
        Message
      </label>
      <textarea
        className=""
        id="contact-message"
        name="message"
        onChange={[Function]}
        required={true}
      />
      <input
        type="submit"
        value="Send"
      />
    </form>
  </div>
</section>
`;
